### DISCUSSION TITLE
the problem with using tauri Rust Commands

### DISCUSSION THREAD

**Discussion by hanxiaohuivc** [tauri-apps/tauri#2495] [Q&A] [answered]
Hi everyone. I encountered a problem when using ‘Rust Commands’. 
The problem is that I pass a parameter from js to rust and return a result. The code is as follows:
function rustwebapi() {
        invoke('jstorustwebapi', {invokeMessage:'https://jsonplaceholder.typicode.com/posts/2' })
        .then((message) => {
            let textId = document.getElementById('#rustwebapi');
            textId.val(JSON.stringify(message));
        })
    }

My rust code is as follows：

async fn get_post_as_string2(url: String) -> Result<String, Box> {
    println!("get_post_as_string2:{}", url);
    let s = reqwest::get(&url).await?.text().await?;
    println!("get_post_as_string2:{}", s);
    Ok(s)
}

async fn some_other_function(invoke_message: String) -> Option {
    let webapistr = get_post_as_string2(invoke_message).await.unwrap();
    Some(webapistr.into())
}

#[tauri::command]
async fn jstorustwebapi(invoke_message: String) -> Result<String, String> {
    let result: Option = some_other_function(invoke_message).await;
    if let Some(message) = result {
        Ok(message.into())
    } else {
        Err("No result".into())
    }
}

But when calling, the following error is returned:
thread'tokio-runtime-worker' panicked at'not currently running on a Tokio 0.2.x runtime.', C:\Users\hanxi\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-0.2.25\ src\runtime\handle.rs:118:28
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

How to solve the problems? Thank you.

**Comment by FabianLars** [comment]
Hey. This error most likely stems from your specific rust code (in your `jstorustwebapi` command) and not from tauri itself. So either your code or one of your dependencies in rust is trying to use a heavily outdated version of tokio.
So check your rust code and dependencies and if you can't find it, please edit your question to show us your `jstorustwebapi` command code in rust.

**Comment by hanxiaohuivc** [comment]
My rust code is as follows, I am in the process of learning, so the rust code may be messy:

async fn get_post_as_string2(url: String) -> Result<String, Box<dyn std::error::Error>> { 
  println!("get_post_as_string2:{}", url); 
  let s = reqwest::get(&url).await?.text().await?;  
  println!("get_post_as_string2:{}", s); 
  Ok(s)
}

async fn some_other_function(invoke_message: String) -> Option<String> {
  let webapistr = get_post_as_string2(invoke_message).await.unwrap(); 
  Some(webapistr.into())
}

#[tauri::command]
async fn jstorustwebapi(invoke_message: String)  -> Result<String, String> {  
  let result: Option<String> = some_other_function(invoke_message).await;
  if let Some(message) = result {
    Ok(message.into())
  } else {
    Err("No result".into())
  }  
}

**Reply by FabianLars** [reply]
What is the version of reqwest (in your cargo.toml)? It needs to be at least 0.11.0 (current is 0.11.4) to support tokio 1.x

**Reply by hanxiaohuivc** [reply]
tokio = { version = "0.2", features = ["full"] }

**Reply by FabianLars** [reply]
i asked for the reqwest version :P But yeah tokio itself needs to be at version 1.x too (1.10 is the latest version i think).

You can also use this to check which dependencies are outdated: https://github.com/kbknapp/cargo-outdated

**Reply by hanxiaohuivc** [reply]
reqwest = { version = "0.10", features = ["json"] }

**Reply by FabianLars** [reply]
replace those with:
```toml
tokio = { version = "1.10", features = ["full"] }
reqwest = { version = "0.11", features = ["json"] }
```
and you should be good to go, if you don't have other dependencies like that :D

**Reply by hanxiaohuivc** [reply]
Hello. According to your guidance, my code can run normally, thank you very much. In addition, I have encountered some problems using tauri in linux, can you please pay attention to it? https://github.com/tauri-apps/tauri/discussions/2498#discussion-3530523

**Comment by hanxiaohuivc** [comment, marked as answer]
@FabianLars According to your guidance, my code can run normally, thank you very much.
FabianLarstokio = { version = "1.10", features = ["full"] }
reqwest = { version = "0.11", features = ["json"] }
